gson/src/main/java/com/google/gson/internal/bind/TypeAdapterRuntimeTypeWrapper.java

```diff
@@ -51,7 +52,7 @@ public void write(JsonWriter out, T value) throws IOException {
 
     TypeAdapter chosen = delegate;
     Type runtimeType = getRuntimeTypeIfMoreSpecific(type, value);
-    if (runtimeType != type) {
+    if (runtimeType != type && !Primitives.areEquivalentTypes(runtimeType, type)) {

```

Comment:
The problem is triggered by primitive vs. wrapper types, but that isn't the root of the problem. The root of the problem is that Gson currently lets the runtime type take precedence over the `@JsonAdapter` annotation. This is weird!

For example, consider a JSON adapter that calls `toString()` on the object and writes that string to the `JsonWriter`:

```
  private static class ToStringAdapter extends TypeAdapter<Object> {
    @Override public Object read(JsonReader in) throws IOException {
      throw new UnsupportedOperationException();
    }

    @Override public void write(JsonWriter out, Object value) throws IOException {
      out.value(value.toString());
    }
  }
```

We can request this adapter like so:

```
  private static final class Box {
    @JsonAdapter(ToStringAdapter.class)
    Object o;
  }
```

But unfortunately due to a mistake in how Gson precedence works, the `ToStringAdapter` is ignored whenever `o`’s type isn’t `Object.class`. For example, this test fails because the `ToStringAdapter` is never used.

```
  public void test() {
    Gson gson = new Gson();

    Box box = new Box();
    box.o = Arrays.asList(5, 6, 7);

    String json = gson.toJson(box);
    assertEquals("{\"o\":\"" + box.o.toString() + "\"}", json);
  }
```

Fixing the precedence is [possible](https://github.com/google/gson/commit/d1770c9f77cd7761add943d81b773b2df4594f3b), but dangerous and potentially incompatible with currently deployed code.